### PR TITLE
 Followup to #17049: change more instances of RuntimeError to IndexError

### PR DESCRIPTION
**Pull request by skrah**

### CODE REVIEW

**Comment by skrah** [comment]
Followup to #17049.

**Review by gchanan — APPROVED** [review]

**Comment by gchanan** [comment]
@skrah it looks like a few tests are failing, mind taking a look?

**Comment by skrah** [comment]
@gchanan I fixed the tests in 7248093; now 70 CIs pass, two have these unrelated failures:

```
data_workers_test.py::DataWorkersTest::testNonParallelModel PASSED [  6%]
Too long with no output (exceeded 1h0m0s)
```

```
t_densenet121
AssertionError: Test model downloading failed
```

File: test/test_torch.py [hunk 1]
```diff
@@ -7986,10 +7986,12 @@ def _test_flip(self, use_cuda=False):
         self.assertRaises(RuntimeError, lambda: data.flip(0, 1, 1))
         # not allow empty list as input
         self.assertRaises(TypeError, lambda: data.flip())
+
         # not allow size of flip dim > total dims
-        self.assertRaises(RuntimeError, lambda: data.flip(0, 1, 2, 3))
+        exc = RuntimeError if use_cuda else IndexError
```
Comment (gchanan) [inline test/test_torch.py]: why does this differ on cpu vs cuda?  Can you fix that?

Comment (skrah) [reply test/test_torch.py]: Yes, this one is in `aten/src/ATen/native/TensorTransformations.h`.  There are other candidate locations for conversion to `IndexError` like for example `aten/src/ATen/TensorOperators.h`:

```
  AT_CHECK(
    index.isIntegral(),
      "Can only index tensors with integral scalars");
```

Would you prefer all of them in one PR?

Sometimes the choice of exception is not obvious, for example in the above `data.flip(0, 1, 2, 3)` case `ValueError` or `TypeError` might be appropriate as well.

On the other hand it's probably easier if users just have to catch a single exception for all cases where "something indexing related" goes wrong.

Comment (gchanan) [reply test/test_torch.py]: I don't think all of them need to be in the same PR, but it's strange if the behavior isn't the same on CPU and CUDA.

**Review by gchanan — APPROVED** [review]

**Comment by skrah** [comment]
@gchanan Sorry, I didn't see the approval and pushed more conversions shortly afterwards.

I simply searched the sources for `AT_CHECK` and converted everything that seemed uncontroversial to me.

**Comment by gchanan** [comment]
@skrah: no apologies necessary, this is a nice change :).  But let's put any more changes in a new PR at this point.

Also, if you are going to add more I'd recommend adding an AT_INDEX_CHECK, I think the code ends up more clear and you don't risk making a mistake with the logic.

**Review by gchanan — APPROVED** [review]

**Comment by skrah** [comment]
Yes, `AT_INDEX_CHECK` sounds like a good idea if I find more locations. Thanks for reviewing again!

**Pull request closed without merging**